### DISCUSSION TITLE
Proposal: Extend generic type new() constraint with parameter types

### DISCUSSION THREAD

**Discussion by devedse** [dotnet/csharplang#769] [General]
_From @devedse on October 12, 2016 20:20_

## Background

Currently C# allows its users to apply the "new()" constraint on methods accepting a generic type to let the compiler know it has a parameter-less constructor.

The following example shows 2 'random generators' that implement the IRandom interface that are being created and then used to return data:

```
class Program
{
    static void Main(string[] args)
    {
        var randomData1 = CreateRandomData<FastRandom>(10);
        var randomData2 = CreateRandomData<CryptoRandom>(10);
    }

    public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, new()
    {
        var randomGenerator = new R();
        for (int i = 0; i < length; i++)
        {
            yield return randomGenerator.Next();
        }
    }
}

public interface IRandom
{
    int Next();
}

public class FastRandom : IRandom
{
    public FastRandom()
    {

    }

    public int Next()
    {
        return 0;
    }
}

public class CryptoRandom : IRandom
{
    public CryptoRandom()
    {

    }

    public int Next()
    {
        return 1;
    }
}
```
## Problem

Currently it's not possible to provide a parameter (e.g. a seed) to the constructor of the 'to-be created' random generators. The current solutions would be:
1. Use reflection to call the constructor and passing the argument to it. (Disadvantage: Not compile time safe)
2. Use a factory and do a switch case on the different types of implementations of the IRandom interface. Then return a new instance of the found match. (Disadvantage: Need to extend Factory every time a new class implements IRandom)
3. Implement an additional method on the interface (and thus implementing classes) Initialize(int seed); which will initialize the class and set it up. (Disadvantage: Having to do 2 method calls for something that's normally done in the constructor.)
## Solution: parameterized new()

My proposed solution would be to also accept parameter types in the new() constraint. E.g. the **new(int)** in the following line of code:

```
public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, new(int)
```

This would make the complete solution look something similar to this:

```
class Program
{
    static void Main(string[] args)
    {
        var randomData1 = CreateRandomData<FastRandom>(10);
        var randomData2 = CreateRandomData<CryptoRandom>(10);
    }

    public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, new(int)
    {
        var randomGenerator = new R(50);
        for (int i = 0; i < length; i++)
        {
            yield return randomGenerator.Next();
        }
    }
}

public interface IRandom
{
    int Next();
}

public class FastRandom : IRandom
{
    private int seed;
    public FastRandom(int seed)
    {
        this.seed = seed;
    }

    public int Next()
    {
        return 0 + seed;
    }
}

public class CryptoRandom : IRandom
{
    private int seed;
    public CryptoRandom(int seed)
    {
        this.seed = seed;
    }

    public int Next()
    {
        return 1 + seed;
    }
}
```

_Copied from original issue: dotnet/roslyn#14461_

**Comment by devedse** [comment]
_From @dsaf on October 12, 2016 21:24_

This seems to be a duplicate of #2206 which is a shame because your version is also well-written. On the upside it's already in `1-Planning` state :).

**Comment by devedse** [comment]
Ah :), well, my goal to get this implemented would still be resolved then.
Any idea how long it usually takes for an item to go from 1-Planning to Implemented / Released?

**Comment by devedse** [comment]
_From @dsaf on October 13, 2016 16:47_

No idea. Let's say I wouldn't expect it next year.

**Comment by devedse** [comment]
_From @timopomer on November 28, 2016 0:55_

@devedse i know this issue is old, but just in case you wanted to make this you can use (R)Activator.CreateInstance(typeof(R), length)

**Comment by devedse** [comment]
Yeah, that's indeed one of the workaround you can use for now. However I would like to do this type safe at compile time :).

**Comment by Mafii** [comment]
Well written Proposal. I would love to see this in the language. More compile time safety like this makes big codebases easier to maintain.

**Comment by orthoxerox** [comment]
Some other potential solutions for this:

1. Abstract factory:

```c#
public interface IRandomFactory
{
    IRandom Create(int length);
}

public struct FastRandomFactory : IRandomFactory //being a struct is important
{
    public IRandom Create(int length) => new FastRandom(length);
}

public static IEnumerable<int> CreateRandomData<F>(int length) 
    where F : IRandomFactory, struct
{
    var randomGenerator = default(F).Create(length);
    for (int i = 0; i < length; i++)
    {
        yield return randomGenerator.Next();
    }
}
```

2. Waiting for shapes (they will solve a generalized problem of requiring specific methods)

3. Extending interface definitions to support constructors:

```c#
public interface IRandom
{
    public IRandom(int length);
    ...
}
```

**Comment by jnm2** [comment]
The thing I'm worried about is that `new(int, string)` has no semantic meaning. Types that aren't meant to match will match. Also suppose you need to refactor to start taking three parameters? All your method constraints are coupled to the very specific signature and all will break, whereas a factory interface or custom delegate are named and separate that coupling away from each usage quite nicely.

**Reply by TahirAhmadov** [reply]
> The thing I'm worried about is that `new(int, string)` has no semantic meaning. 

I can see how that's a problem with delegates - you can assign any method to the delegate, even if it's not a semantic fit. However, in this case, we're talking about constructors only; the context is limited enough - the semantics are "create new object".

> Types that aren't meant to match will match. 

That's very unlikely to happen; you use base class/interface constraints, as needed, to narrow the pool of potential generics. It's important to remember that this constraint is not meant to restrict types, rather it's meant to allow a type safe way to create objects.

> Also suppose you need to refactor to start taking three parameters? All your method constraints are coupled to the very specific signature and all will break, whereas a factory interface or custom delegate are named and separate that coupling away from each usage quite nicely.

The same number of breaks and required fixes will ensue with delegates and factories. Fundamentally, you used to call a method with 2 arguments; and now need to change it to call 3 arguments. Whether that method is a delegate, a factory, or a constructor, doesn't change the amount of work of making this change. Furthermore, it's even more work - now you have to write and maintain the delegate/factory stuff, in addition to the constructor.

**Comment by Pzixel** [comment]
Allowing interfaces to have static methods, or add new interfaces on types themselves could solve this. Create an instance of class just to call a method which returns new instance of this class. That's sounds weird. This is why we cannot have interface `IParseable<T>` where we could have a method `T Parse(string input);`, because we don't want to create an instance just to create an instance and simultaneosly we can't invoke `T.Parse` because we don't constraint the type itself, but only its instance methods...

I know about implementation of generics with __Cannon so I don't have a good solution of how it could be implemented. However, we probably might add costructors to interfaces so I'd be:
```csharp
public interface ICreatable
{
    string GetName();
    new(string name);
}
```

**Reply by RikkiGibson** [reply]
Interfaces now can have static abstract methods :)

**Comment by jnm2** [comment]
@Pzixel Already gave my thoughts on that in https://github.com/dotnet/csharplang/issues/837.

**Comment by Pzixel** [comment]
@jnm2 shapes probably might solve this? I mean:

```csharp
public shape CreatableWithSeed
{
    new(int seed);
}

public static IEnumerable<int> CreateRandomData<R>(int length) 
   where R : IRandom, CreatableWithSeed
{
    var randomGenerator = new R(50);
    for (int i = 0; i < length; i++)
    {
        yield return randomGenerator.Next();
    }
}
```

**Comment by gafter** [comment]
This is part of a proposal being tracked at #110 which is under active development at MSR (Microsoft Research).

**Comment by ActivistInvestor** [comment]
> @jnm2: The thing I'm worried about is that new(int, string) has no semantic meaning

I don't recall ever having used new() without an additional type constraint. 

How does using new() without a type constraint have any more semantic meaning than using the theoretical or proposed parameterized new(...) without also using a type constraint?

**Comment by jnm2** [comment]
> I don't recall ever having used new() without an additional type constraint.

It's common as part of a generic factory method (in the same spirit as `new Lazy<Foo>().Value`).

> How does using new() without a type constraint have any more semantic meaning than using the theoretical or proposed parameterized new(...) without also using a type constraint?

This is an excellent point. I would answer that it's obvious with `new()` that you get the default construction of whatever `T` is, whether constrained or not. With `new(string, int)`, it's not obvious what the first and second arguments mean, whether `T` is constrained or not. What choices are you making with that string and int? What are you causing to happen?

You don't really have that question with `new T()`. You know you're getting the default construction of `T`.

**Comment by ActivistInvestor** [comment]
> This is an excellent point. I would answer that it's obvious with new() that you get the default construction of whatever T is, whether constrained or not. With new(string, int), it's not obvious what the first and second arguments mean, whether T is constrained or not. What choices are you making with that string and int? What are you causing to happen? You don't really have that question with new T(). You know you're getting the default construction of T.

I don't see how that differs from what one must do today (use reflection to construct instances via calls to a parameterized constructor).

This is widely done now, except that the construction is done via reflection, and the consumer of the class that performs the construction has to understand what the arguments of the constructor they have to provide mean. The only difference between that, and a parameterized new() constraint, is that it becomes a compile-time error rather than a runtime error.

So, are you suggesting that what we must currently do today, we shouldn't do, because the meaning of the constructor's arguments aren't any more obvious than that of a parameterized new() constraint?

One way or the other, it requires documentation.

**Comment by jnm2** [comment]
> I don't see how that differs from what one must do today (use reflection to construct instances via calls to a parameterized constructor).
> 
> This is widely done now

Is it widely done now? Other than serializers whose entire job is reflection-based, I don't see this much at all, and I have been in a variety of codebases. Activator.CreateInstance is very rarely used with constructor parameters- certainly no more than `MethodInfo.Invoke` and for the same reasons.

> what one must do today

What people usually do is a factory delegate or interface where you can call `IPersonFactory.Get(string name, int age)` and that is much more common. Even a `Func<string, int, Person> getPersonByNameAndAge` parameter has all the semantic meaning you need. No Activator.CreateInstance nonsense.

**Comment by ActivistInvestor** [comment]
> Is it widely done now? Other than serializers whose entire job is reflection-based, I don't see this much at all, and I have been in a variety of codebases. Activator.CreateInstance is very rarely used with constructor parameters- certainly no more than MethodInfo.Invoke and for the same reasons.

It is done via ConstructorInfo, not Activator. The patterns it is used in typically involve a class that has the responsibility for managing multiple instances of a consumer-provided class (the generic argument), and the class that manages the instances must create instances of the consumer-provided class, and pass arguments to their constructors. In many cases, the consumer-provided class is derived from another class that must have the constructor arguments passed into their constructors, etc.

The practice has been discussed in this issue, nonetheless.

> What people usually do is a factory delegate or interface where you can call IPersonFactory.Get(string name, int age) and that is much more common. Even a Func<string, int, Person> getPersonByNameAndAge parameter has all the semantic meaning you need. No Activator.CreateInstance nonsense.

In many of the cases where I've dealt with this pattern, there is no need for the consumer to provide a factory function, make any API calls, or do anything at all in code. 

Everything is done via an Attribute which is read when the containing assembly is loaded. The consumer only needs to apply the attribute to the class for which instances are created and managed, and ensure that the class has the required constructor. When the assembly containing the attributed class is loaded, the system that manages instances of the class will register the class and create instances as needed, without the consumer that provides the class having to do anything at all.

I've used patterns that use factory functions a la Lazy<T> as well, but in the scenario I describe above, the design goals were to minimize the burden placed on consumers that provide the classes that are to be managed.

**Comment by ActivistInvestor** [comment]
> Is it widely done now? Other than serializers whose entire job is reflection-based, I don't see this much at all, and I have been in a variety of codebases. Activator.CreateInstance is very rarely used with constructor parameters- certainly no more than MethodInfo.Invoke and for the same reasons.

Actually, most robust factory patterns do not have a dependence on generics in the first place, because if they did, that would restrict them to only being usable with statically-defined types, rather than types that are defined or determined dynamically at runtime (which is certainly not the exclusive domain of de-serializers). Robust factory patterns may support support generics at some level, but when you get down to the bare metal, the generic arguments usually get converted to a runtime type via `typeof()`, and that's what's ultimately used to construct instances, which can only be done via reflection.

For example, if there was only `Activator.CreateInstance<T>`, it wouldn't be terribly useful if you don't know the type at compile-time.

**Comment by devedse** [comment]
The reason I'm not sure if I would like Activator.CreateInstance<T> is because it would make this functionality rely on the System.Reflection library. I suppose it's better then nothing, but for a feature like this I'm not sure if that is the way to go. Furthermore, would there be any difference performance wise? (When extending Activator.CreateInstance I would assume reflection would be used in combination with casting, wheras the new(... , ...) could be as fast as simply creating an instance of a class.

**Comment by Joe4evr** [comment]
> The reason I'm not sure if I would like Activator.CreateInstance is because it would make this functionality rely on the System.Reflection library.

Well, using the `T : new()` constraint and doing `new T()` is *already* using `Activator.CreateInstance()` since forever, so you're out of luck on that front. 🙃 

> When extending Activator.CreateInstance I would assume reflection would be used in combination with casting, wheras the new(... , ...) could be as fast as simply creating an instance of a class.

There is no way to invoke the constructor of a type known only at runtime (which is exactly what you're doing) without using reflection.

**Comment by devedse** [comment]
@Joe4evr, I don't think your second point is fully correct. Since you can use Expression trees to in fact reach the speed of Compile time invoked constructors.

**Comment by devedse** [comment]
HI all,

Can we maybe revisit this and get some new insights if we would want or not want this to be implemented, as there's quite some people that have upvoted this.

**Comment by jnm2** [comment]
I really like the way https://github.com/dotnet/csharplang/issues/2528 would allow this.

**Reply by pzahra** [reply]
If it were a matter of using an interface (or variety thereof) to enforce a specific constructor signature (even if it has to be explicitly implemented), then you don't need a constructor constraint and can use the one in the interface.

Otherwise the callback method works just fine.

**Comment by Shadowblitz16** [comment]
I actually just tried doing `where T : new(...)` just to see if it was a thing.
then I looked up this when I realized C# didn't support it.
it kinda makes sense syntax wise to allow the user to control generic constructors by passing parameters to the new constraint.

its basically saying constrain the constructors to this signature.
but I prefer this idea over shapes since I haven't learned those. +1 

EDIT: what could be done is allowing a interface to define a constructor template. like so.. 
```
using System.Collections.Generic;
using System.Text;

namespace FarCore.Model
{
    public interface IEntity : IUpdatable
    {
        string         Name { get; }
        IEntityConfig Config { get; }

        IEntity(string name, IEntityConfig config); //here
    }
}
```

and make the new constraint either require the same parameters or just auto detect the paramaters in the IEntity constructor.

**Comment by Pzixel** [comment]
with shapes it should be possible to demand static method Create which should actually create a value.

**Comment by darssy** [comment]
> HI all,
> 
> Can we maybe revisit this and get some new insights if we would want or not want this to be implemented, as there's quite some people that have upvoted this.

@devedse considering how popular is this [10 year old question at stackoverflow](https://stackoverflow.com/q/840261/1303323), I would say it's not just "some people" ;)

I personally consider it useful as I found myself more than once in need of a parameterized new() solution. My 2 cents.

**Comment by devedse** [comment]
Hi all, It's been a while, but I just saw a notification drop in my mailbox so I thought I'd chip in again too. I still think this idea would be great to add.

As @Shadowblitz16 also mentioned, having this, simply makes sense. Having `new()` accept a parameter specification feels intuitive to me, especially since there's already ( ) in there.

I saw your approach regarding shapes as well but I think this could still be a strong addition since shapes seem to resolve a more difficult challenge we're running into.

In regards to @jnm2 's comment about it not having semantic meaning, can't we build it the same way that named tuple's solve this?:

```
private static (int Count, double Sum, double SumOfSquares) ComputeSumAndSumOfSquares(IEnumerable<double> sequence)
{
    var computation = (count: 0, sum: 0.0, sumOfSquares: 0.0);

    foreach (var item in sequence)
    {
        computation.count++;
        computation.sum += item;
        computation.sumOfSquares += item * item;
    }

    return computation;
}
```

That way you could have either one of these options:

Option 1:
```
    public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, new(int)
    {
      ...
    }
```

Option 2:
```
    public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, new(int seed)
    {
      ...
    }
```

Another way I wouldn't mind this issue to be solved is by having the option to define a constructor definition inside an interface. I mean, why do we have the ability to define methods but not constructors? To me it would make sense to have both. This might also have some more benefits when doing serialization / deserialization.

Option 3:
```
public interface IRandom
{
  IRandom(int seed);

  int Next();
}

public class Random : IRandom
{
  public Random(int seed)
  {
    ...
  }

  public int Next()
  {
    ...
  }
}
```

In my opinion, option 3 more closely follows the current C# spec.

I think either one of these options could be implemented in a way that would avoid the use of `Activator.CreateInstance<T>()`

Edit:
I have seen some backlash against Option 3 as people have stated that the constructor definition in the interface doesn't have anything to do with the implementing class.  I do think that after reading through these that it would make more sense to go for Option 1 or 2. To me a constructor feels the same as a method in this sense, but since within C# we've made the decision to put differentiating logic to construct a specific instance of a class within the constructor, having an interface definition of this constructor wouldn't make sense.

Reference: https://github.com/dotnet/csharplang/issues/837#issuecomment-323710951 and https://github.com/dotnet/csharplang/issues/837#issuecomment-323711462

**Comment by Pzixel** [comment]
It could be naturally solved if typeclasses come into the game so you can just write `where T : New<int>`, and there is no need for special syntax in the language. Waiting for #164

**Comment by devedse** [comment]
@Pzixel then you'd still have the problem that it doesn't have any semantic meaning though.

**Comment by Pzixel** [comment]
I don't get your idea. Could you elaborate? It looks like it's a complete replacement:

```cs
shape New<T> {
  static Create(T arg) -> this;
}

...

void Foo<TVal, TNew>(Tval arg) where TNew : New<TVal>, IBar {
  var newVal = TNew.Create(arg);
  newVal.Bar();
}
```

**Comment by devedse** [comment]
Ah that's what you mean, I haven't really checked out shapes enough. So with that solution you wouldn't be using the constructor, but rather a static method that creates an instance of the class with some parameters?

Could you maybe show an example how this would work for the example above with the Random instances?

**Comment by Pzixel** [comment]
No problem:

```cs
public extension NewRandom of Random : New<int>
{
    public static this Create(int arg) => new Random(arg);
}

public static IEnumerable<int> CreateRandomData<R>(int length) where R : IRandom, New<int>
{
  var randomGenerator = R.Create(50);
  for (int i = 0; i < length; i++)
  {
    yield return randomGenerator.Next();
  }
}
```

I'm also using a quite enhanced version of `this` proposal: #169 , but I think it's clear what do I mean.

**Comment by devedse** [comment]
@Pzixel, but in your example, why do you still include the `: New<int>`. You already specify in the `Create` method that there's an `int` argument right?

**Comment by Pzixel** [comment]
Because type inference in C# works this way (or to be more precise: it doesn't).

**Comment by darssy** [comment]
@devedse as I'm the one who "stirred the pot" I think that it might be better off if we keep the constructor in interface separate from the parametarized `new()` feature request. Otherwise it won't have a chance to make it to the language I'm afraid.

Now the question (IMHO always) is how can we draw some attention to this one so that it will have some chance to reach a future version of C#. C# 20 let's say -hope I wont be in pension by then.😅 As I see it, the popularity of the question in SO (where I actually landed this page from) can be an indicator of the usability of the feature.

**Comment by dogac00** [comment]
Any updates on when could this be available in C#?

**Reply by jnm2** [reply]
I think the general rule is that the answer to this question is always no.

**Comment by pzahra** [comment]
Let's step back a minute.

The goal we're trying to achieve here is simple: at some point within the body of this function, we _may_ want to create a new instance of `IRandom` (providing it an implementing classname) and pass it a Seed int. That's all.

Here's the original function using a callback instead:
`IEnumerable<int> CreateRandomData(CreateIRandomCallback create, int length)`

`delegate IRandom CreateIRandomCallback(int seed)`

(of course you could use `Func<int, IRandom> create` instead of declaring a delegate type, but that's another discussion)

Usage:
`values = CreateRandomData(seed => new CustomRandom(seed), 100);`

Note that the end result _doesn't even need to be generic._ The function still has the ability to create a new instance of `IRandom` wherever/whenever it wants, using the signature it is expecting. On the other side, we're still providing it a classname, but we're also choosing the constructor. Maybe it's one that has additional default params we're not using; maybe we're specifying Seed as a property, or ignoring it entirely; maybe we're providing it with upper or lower bounds, or a tablename to query. Whatever the case: it gave us a Seed, we gave it an IRandom, and it continued doing whatever it does.

This is _more flexible_ than the proposed generic constraint, doesn't require every implementer of IRandom to have a potentially breaking constructor, doesn't make use of reflection, doesn't require you to be the owner of every implementer, and is already fully supported.

**Comment by MithrilMan** [comment]
Ok it seems I expressed the same need https://github.com/dotnet/runtime/issues/68858

**Comment by jnm2** [comment]
If the language had `this` types, maybe that in combination with static abstract methods would fill the need and would allow a nominal rather than shape-based approach:

```cs
public void M<T>() where T : IFoo // instead of where T : new(...), now you have something with a specific meaning
{
    T instance = T.Create(someSemanticInfo: 42);
    // ...
}

// Having to declare the signature this way means that not all `new(int)`-shaped constructors are silently interchangeable
// regardless of what the parameters mean.
public interface IFoo
{
    public static abstract this Create(int someSemanticInfo);
}
```

**Reply by pzahra** [reply]
Adding statics to interfaces would be neat for other situations, but we can already answer the original question of this topic with a callback.

**Reply by HaloFour** [reply]
It might be interesting if an interface could describe a constructor as a required static member instead of a factory method.

**Reply by jnm2** [reply]
> Adding statics to interfaces would be neat for other situations, but we can already answer the original question of this topic with a callback.

ℹ Adding static abstract methods to interfaces *is* neat and has already shipped in C# 10 under a preview flag.
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/static-abstract-interface-methods

**Reply by MithrilMan** [reply]
This wouldn't solve the problem if you can't alter the generic class you want to instantiate (you aren't the owner) declaring that the generic type implements that interface

**Reply by HaloFour** [reply]
That's where [Roles and Extensions](https://github.com/dotnet/csharplang/discussions/5498) would come in.

**Reply by jnm2** [reply]
@MithrilMan That also seems like a less likely situation to be in. It verges on shape-based matching which personally would bother me.

**Reply by MithrilMan** [reply]
> @MithrilMan That also seems like a less likely situation to be in. It verges on shape-based matching which personally would bother me.

@jnm2 
I've shown an example on my issue here https://github.com/dotnet/runtime/issues/68858

**Reply by jnm2** [reply]
@MithrilMan So yes, that example bothers me. What if the exception class passed as T has a constructor with a single string parameter and the string parameter is *not* `message`? I've created such exception classes.

An example of such a thing in the BCL is `ArgumentNullException`. It's semantically wrong to feed an exception message sentence into the single string parameter without knowing how it will be used.

**Reply by TahirAhmadov** [reply]
@jnm2 how is that different than calling `IExceptionFactory.Create(string message)` like `Create(nameof(someParam))`? There is no escaping the ubiquity of `string` (at least, fully - you can use `Uri` and `FileInfo` stuff to an extent).

**Reply by jnm2** [reply]
@TahirAhmadov It's different because the exception class would have to implement `IExceptionFactory.Create(string message)` and do something semantically correct with a variable named `message`. (Or, with roles and extensions, you could provide this implementation and handling of `message` externally.) There's no accidental passing of `string message` to `string paramName`.

**Reply by TahirAhmadov** [reply]
It's kind of similar, though. Option A - you have to make sure to use an `Exception` type as generic argument which has a ctor accepting a single `message` parameter. Option B - you have to make sure to pass a message to a `message` parameter and not `paramName`. In both cases, it's entirely up to the developer to ensure the correct thing is called. 
Yes, you have a little more visual feedback - the IDE will show you the parameter name in one case and will not do that in another. However, that visual feedback does not guarantee anything. It's still very possible to accidentally pass a `message` to a `paramName` or vice-versa.

**Reply by jnm2** [reply]
@TahirAhmadov It's not implicit though. If you make a mistake in implementing `IExceptionFactory` for `ArgumentNullException`/whatever exception, you are explicitly passing `message` to `paramName` within the responsible exception class. If there is only a `where T : new(string)` constraint, then there is an implicit bug not located within the *implementation* of `IExceptionFactory` (Within the responsible exception class, or with roles and extensions, within code that has taken responsibility for this implementation.)

**Reply by MithrilMan** [reply]
Whie I agree on your concern about having a specialized Exception type with a constructor (string) that change its meaning (message <-> paramName), that example could be a convenient method based on convention (the convetion in this case is that your string parameter is a message and when you use that method, you feed it with an Exception that adhere to such convention).

If there are exceptions that doesn't adhere to the convetion, it could be a dev effort to use that string parameter based on the implementation of the exception he wants to rise. I could rename that variable as a generic `string argument` so that the developer has to think about how the argument is get used in the Exception constructor itself.

That simple example at the end is just a way to have a factory for generic type with constructor arguments without relying on reflaction and Activator.CreateInstance, it doesn't mean the dev doesn't have to know what to pass to the method (we can argue about it being bad code and I'd agree, but would be OT)

**Reply by jnm2** [reply]
I see. If you did rename it `string constructorArgument` or something like that, then I'm no longer concerned. But the usefulness of this helper seems limited. If the argument(s) are just going to be passed through from the call site untouched, why not inline the `if` statement and the exception creation and throwing? The latter is far more flexible, too, allowing every combination of counts and types of arguments to be succinctly specified. You get intellisense, etc.

(Alternatively, https://github.com/safakgur/guard is pretty cool.)

**Reply by MithrilMan** [reply]
The point here is not about how to implement a Guard (I wouldn't use the example code in production as is), the concern here is about how to prevent to make use reflection and improve performance when there is the need to instantiate a generic with a constructor with parameters. We could argue about if the pictured scenario happens often or not but the code is out of scope.

Thanks anyway for the repo link, tbh I see the suggested library is fluent so it means it has always to allocate an object instance to be fluent, I'd prefer something more close to this instead https://github.com/ardalis/GuardClauses

**Reply by jnm2** [reply]
There will be the exact same reflection either way. `where T : new()` is implemented via `Activator.CreateInstance<T>`.

There are no allocations. From its readme:

> If this looks like too much allocations to you, fear not. The arguments are read-only structs that are passed by reference. See the [design decisions](https://github.com/safakgur/guard#design-decisions) for details and an introduction to Guard's more advanced features.

**Reply by TahirAhmadov** [reply]
> It's not implicit though. If you make a mistake in implementing `IExceptionFactory` for `ArgumentNullException`/whatever exception, you are explicitly passing `message` to `paramName` within the responsible exception class. If there is only a `where T : new(string)` constraint, then there is an implicit bug not located within the _implementation_ of `IExceptionFactory` (Within the responsible exception class, or with roles and extensions, within code that has taken responsibility for this implementation.)

But you do explicitly use `ArgumentNullException` as the generic argument. I agree - there is not enough "visual feedback" - it's easier to make a mistake; however you know your type has a constructor which does not do what the receiving generic type expects.
Also, I don't see this as a frequent situation which would lead to numerous bugs. It's much more common to have domain specific types which don't have these kinds of constructor collisions. In fact, `Argument***Exception` is a notorious exception (pun intended) to the expectation of constructor parameters in BCL and definitely elsewhere in the C# ecosystem.
More importantly, though, even with this one minor caveat of `Argument***Exception`s, this is still a better approach than either a) using `Activator.CreateInstance` directly, which is much more error prone, or b) writing a new factory interface and N types - one for each possible generic argument type; lots of make work.

> There will be the exact same reflection either way. `where T : new()` is implemented via `Activator.CreateInstance<T>`.

That's a separate conversation. I don't know if this is optimized down the line or not (say, by JITter). Personally, I think it's due time to vastly improve this even in C# - it's easy enough to write IL which binds to the constructor as part of generic type initialization. Personally, I use a helper class which uses IL emit instead of `CreateInstance` which does just that - to be on the safe side.

**Reply by MithrilMan** [reply]
> There will be the exact same reflection either way. where T : new() is implemented via Activator.CreateInstance<T>.

At first I didn't belived it, I double checked on sharplab and I confirm what you say, I don't know why it doesn't have a better code to bind to the default constructor, hopefully it will be improved when the eventual `new(string)` will be a thing.

> There are no allocations. From its readme:
> > If this looks like too much allocations to you, fear not. The arguments are read-only structs that are passed by reference. See the [design decisions](https://github.com/safakgur/guard#design-decisions) for details and an introduction to Guard's more advanced features.

Even if it's in the stack, it's anyway an allocation, it creates an object it has to be passed along the chain of fluent methods, spends CPU time and since I'd use Guards mostly for basic argument validation (null check, emtpy string, etc...) and they are used in hot paths, that performance penality matters (at least in performance critical services)

If I have to do domain entities validation and readability gained by a fluent syntax helps, then I'm fine using libraries like this or FluentValidation, anyway I'm going off track talking about guard and at the end there is no silver bullet anyway, it's a matter of context (the app type, if it's performance critical, the team skill/background? etc...)

Thanks for the convo

**Comment by koliyo** [comment]
Here is a factory example using the static abstract interface methods in C# 11, that seems to be a useful workaround, continued from what @jnm2 mentioned.

```cs
public interface IFactory<TSelf> where TSelf : IFactory<TSelf> {
  static abstract TSelf New(int i, string s);
}

public struct Foo : IFactory<Foo> {

  public static Foo New(int i, string s) {
    return new Foo(i, s);
  }

  public readonly int I;
  public readonly string S;

  public Foo(int i, string s) {
    I = i;
    S = s;
  }
}

public static class Maker {
  public static T Make<T>(int i, string s) where T : IFactory<T> {
    return T.New(i, s);
  }
}
```

But as @MithrilMan mentions, this only works if you are the owner of the target class.
